pkg/daemon/multus/resources.go

```diff
@@ -0,0 +1,290 @@
+/*
+Copyright 2023 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multus
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	core "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type podNetworkInfo struct {
+	// node the pod is running on
+	nodeName string
+
+	// multus public addr attached (if any)
+	publicAddr string
+
+	// multus cluster addr attached (if any)
+	clusterAddr string
+}
+
+var manualCleanupInstructions = fmt.Sprintf(
+	"manually delete owner configmap %q, and wait for all multus-validation-test resources to be deleted", ownerConfigMapName)
+
+var previousTestSuggestion = "there could be a past test preventing this one from proceeding; " + manualCleanupInstructions
+
+var unableToProvideAddressSuggestions = []string{
+	"multus may be unable to provide addresses for pods",
+	"check networking events on the pod and multus logs",
+	"macvlan: NIC or switch hardware/software may block the association of some number of additional MAC addresses on an interface",
+	"macvlan: interfaces and network switching must enable promiscuous mode to allow receiving packets for unknown (Multus) MACs",
+	"macvlan/ipvlan: switch hardware/software may block an interface from receiving packets to an unknown (Multus) IP",
+}
+
+// create a validation test config object that stores the configuration of the running validation
+// test. this object serves as the owner of all associated test objects. when this object is
+// deleted, all validation test objects should also be deleted, effectively cleaning up all
+// components of this test.
+func (vt *ValidationTest) createOwningConfigMap(ctx context.Context) ([]meta.OwnerReference, error) {
+	testConfig, err := json.Marshal(*vt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to render validation test config [%+v] to a string: %w", *vt, err)
+	}
+
+	isImmutable := true // don't let users modify the test config after it starts
+	c := core.ConfigMap{
+		ObjectMeta: meta.ObjectMeta{
+			Name: ownerConfigMapName,
+		},
+		Immutable: &isImmutable,
+		Data: map[string]string{
+			ownerConfigMapConfigKey: string(testConfig),
```

Comment:
Why do we need to store the test config in a configmap? Do we ever deserialize it? If the test fails and needs to restart, wouldn't we just create a fresh config? Or is there a need to restart after a failed test?

Reply:
We don't. I suppose it could be a configmap with nothing in it. That just felt strange. Its primary purpose is to be the "owner" of all the other test resources so users can clean up the test just by deleting the configmap -- and same for the `cleanup` command.

Reply:
I'll change that.